NIFI-2892 Implement AWS Kinesis Stream Get Processor

ChrisSamo632 opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Enables fetching data from AWS Kinesis Data Streams.

Builds upon the [original PR](https://github.com/apache/nifi/pull/3582) from @udaykale (and comments there by @turcsanyip and @jaynpearl).

Based largely upon the [AWS Kinesis Consumer](https://github.com/aws/aws-sdk-java/tree/master/src/samples/AmazonKinesis) example classes.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- ~[ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?~
- [x] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- ~[ ] Have you ensured that format looks appropriate for the output in which it is rendered?~

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

bhaveshpatelh commented:
Thanks! Looking forward to this piece going out

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/GetKinesisStream.java:
```diff
@@ -0,0 +1,474 @@
+                        .withCommonClientConfig(getClient().getClientConfiguration());
+
+        kinesisClientLibConfiguration.withRegionName(getRegion().getName())
+                .withMaxInitializationAttempts(numRetries);
```
Would be better to allow use of Dynamic Properties to set this and other Kinesis Client Library Configuration properties. There are a lot of KCL config settings that could be tweaked and allowing a user to set them (without increasing the processor's already rather large list of static properties) would seem sensible.

I'll look to push a change for this in the near future.

ChrisSamo632 replied:
Second commit includes the update to allow dynamic property use for configuring the KCL

auyer commented:
Hi ! I'm interested in testing this out.
I currently run Apache Nifi using the base Docker images. How can I proceed to include the Kinesis Get Processor into my images? 
Thanks ! Hope I can help!

ChrisSamo632 commented:
> Hi ! I'm interested in testing this out.
> I currently run Apache Nifi using the base Docker images. How can I proceed to include the Kinesis Get Processor into my images?
> Thanks ! Hope I can help!

@auyer I think you'd need to download the source from my branch, [build it locally](https://github.com/apache/nifi#getting-started) and then copy the new NARs into your image in the `lib/` folder (overwrite the existing AWS NARs) - you can probably build just the AWS NARs (rather than the whole of NiFi) to speed up the process... if you build the whole of NiFi, you might as well just run the compiled version natively rather than using the Docker Image.

So, something like:
```bash
# clone the code
git clone https://github.com/ChrisSamo632/nifi.git
git checkout NIFI-2892

# build the code
cd nifi/nifi-nar-bundles/nifi-aws-bundle
mvn -T 2.0C clean install

# get the NARs
find . -name "*.nar" -exec {} /tmp
```

Copy these into a custom Docker Image with something like the following:
```dockerfile
FROM apache/nifi:1.13.0

RUN rm lib/nifi-aws*.nar

COPY [ "*.nar", "lib/" ]
```

Then run the custom image in the way you normally would.

Note that I've not tested the above, so you may need to correct the instructions as you go - when developing NiFi, I just build from source and run natively, I've not got to the point of trying to copy the into a Docker container yet although that is how I run NiFi normally.

auyer commented:
I've been testing this for a few days, and had no issues so far. I should say that I'm not running NiFi in production yet, but plan to do so soon, and this feature helps a lot !
If anyone else wants to try it in Docker, I've built a image following the instructions above, and published it publicly in [DockeHub](https://hub.docker.com/repository/docker/auyer/apache-nifi-with-get-kinesis). 

Note: I will remove this repo after this PR gets into the next release.

bhaveshpatelh commented:
@ChrisSamo632 Any timelines for merging/releasing it out?
We have a use-case and wanted to use it in NiFi production.

ChrisSamo632 commented:
> @ChrisSamo632 Any timelines for merging/releasing it out?
> We have a use-case and wanted to use it in NiFi production.

It's ready to go as far as I'm concerned, but it needs someone to review and then a committer to merge it (there's a big backlog of PRs, I've no idea when anyone will get to this... much as I'd like it in too because we want to use it in production)

pvillard31 commented:
@ChrisSamo632 - out of curiosity, did you consider the addition of a record reader/writer like with ConsumeKafkaRecord processors? At the moment, as far as I can tell, we would have one record per flow file. Using the records abstraction would provide options such as schema validation, format transformation, as well as having multiple records in one flow file (which is greatly improving the performances in case there is a high number of messages per second).

ChrisSamo632 commented:
> @ChrisSamo632 - out of curiosity, did you consider the addition of a record reader/writer like with ConsumeKafkaRecord processors? At the moment, as far as I can tell, we would have one record per flow file. Using the records abstraction would provide options such as schema validation, format transformation, as well as having multiple records in one flow file (which is greatly improving the performances in case there is a high number of messages per second).

Think I considered this in our original slack conversation, but thought I'd leave it record writers at the minute for simplicity and to better understand how the Kinesis message (vs. Record) structure works

With the KCL worker and multi-threaded consumer approach taken by the KCL library, we'd need a way of combining the records in the processor too (how do we combine records from different consumers? How about where the consumers are reading from different shards?)

So I thought a record reader may be a sensible extension (fully agree it would be good to include from a nifi perspective) once there's more understanding of how people use the processor and how it works with different kinesis setups (so far I've only really tested it with simple streams)

turcsanyip commented:
@ChrisSamo632 Thanks for picking up this task and implementing this non-trivial processor!

I did not review the code in detail (yet) but spotted some issues regarding thread handling:

The `onTrigger()` method should not block its thread. In NiFi, there is a shared thread pool for running all processors' `onTrigger()` methods. If some processors hold up execution threads, it can lead to starvation of other processors.
I think a similar pattern could be used as in case of `ConsumeAzureEventHub`: start up the Worker at the first execution of `onTrigger()` and then just yield in the subsequent calls.

I believe one Worker per processor should be enough and it is not necessary to maintain a pool of workers. A single Worker can run multiple threads for executing the Kinesis RecordProcessors. As far as I saw, the Worker spins up a thread for each assigned shard by default (RecordProcessor-xxxx threads). So the parallel processing is provided with one Worker too. The code is simpler in this way and there is less overhead at runtime (each Worker has its own "maintenance" threads like LeaseRenewer-xxxx, LeaseCoordinator-xxxx).

The processor cannot stop cleanly due to a bug in KCL. Interestingly, it has just been reported to AWS by someone else: https://github.com/awslabs/amazon-kinesis-client/issues/796
When the processor shuts down the Worker, the Worker leaves behind a thread running. Stopping/starting the processor multiple time would lead to thread leaking. Furthermore, the "zombie" thread(s) prevent NiFi to shut down properly:
`2021-03-26 12:12:52,715 WARN [main] org.apache.nifi.bootstrap.Command NiFi has not finished shutting down after 20 seconds. Killing process.`
For this reason, I think we need to downgrade the KCL version to 1.13.3 until the bug is fixed.

And an idea: the way the processor receives data via KCL is quite similar to other Consume*** processors (like the already mentioned `ConsumeAzureEventHub` or `ConsumeAMQP`), so I would consider to name the processor `ConsumeKinesisStream` instead of `Get~` (actually there is a Get-like/polling API for Kinesis too and another processor may be implemented later using that API).

ChrisSamo632 commented:
@turcsanyip thanks for the comments. I'd been a bit confused/concerned about the thread handling and tried (but failed) to find a suitable example, what you suggest looks like a sensible approach on first inspection - I'll try to make the changes sooner than later

Good spot with the KCL issue, will try the downgrade as you suggest (unless a fix comes from AWS in the meantime)

@pvillard31 I'll look again at using Record Reader/Writer at the same time in a fashion similar to the processors suggested above (although I may still suggest it as a future improvement ticket)

ChrisSamo632 commented:
@turcsanyip started to have a look at making these changes. Using KCL 1.13.3 instead of 1.14.x appears to be straight forward (i.e. no big API changes).

The refactoring to use a single Worker per processor and yield once the Worker has been setup looks fairly straight forward and I agree a simpler implementation *except* that if I am to follow the `ConsumeAzureEventHub` approach and use the `ProcessSessionFactory` version of `onTrigger`, I have to extend the `AbstractSessionFactoryProcessor` base class, but the existing AWS processors have a fairly long chain that extends from `AbstractProcessor` - I don't want to unmarry this processor from all the existing AWS processors and moving all AWS processors to use the different base class feels like a big change and not something we'd want to do.

Is there a straight forward way around this do you think? OIne concern I had with the original implementation was the fact that I was holding on to a single `ProcessSession` and comitting it multiple times (i.e. after every set of messages had been processed by the KCL RecordProcessor) - using the ProcessSessionFactory approach to create a new session every time a new set of Kinesis messages are received would seem better... aside for the above issue of the different abstract processors.

I'll look again at this further, but any guidance is welcome!


@pvillard31 having looked again at the Record processing now I've the Azure processor to compare with, I think includiung them here should be fairly straight forward (famous last words).

ChrisSamo632 commented:
@turcsanyip updated per your suggestions.

Note that this includes:
- added (optional) Record Reader/Writer (@pvillard31 FYI)
- better Thread/Session handling for the use of Kinesis Client Library
- reverted to KCL verison 1.13.3 due to [amazon-kinesis-client#796](https://github.com/awslabs/amazon-kinesis-client/issues/796)
- removed double-validation of dynamic properties (introduced by NIFI-8266 @mattyb149 FYI in case you see an issue with my change to `AbstractConfigurableComponent`)
- change to `AbstractProcessor` to remove the `final` modifier of the `onTrigger` method that takes the `ProcessorSessionFactory` - by default processors extending this class will still need to implement the `onTrigger` method that takes the `ProcessSession`, but can now optionally override the former (this seemed a better option in the end than copying the contents of `AbstractProcessor` into `AbstractAWSProcessor`, but happy for that to be challenged)

bhaveshpatelh commented:
@ChrisSamo632 When I tried following commands to build the processor, it's throwing an error.
can you guide me with the next steps?

```
# clone the code
git clone https://github.com/ChrisSamo632/nifi.git
git checkout NIFI-2892

# build the code
cd nifi/nifi-nar-bundles/nifi-aws-bundle
mvn -T 2.0C clean install
```

```
[INFO] -------------------------------------------------------------
[WARNING] COMPILATION WARNING : 
[INFO] -------------------------------------------------------------
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/credentials/provider/factory/strategies/AnonymousCredentialsStrategy.java:[26,29] [deprecation] StaticCredentialsProvider in com.amazonaws.internal has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideCEncryptionStrategy.java:[22,32] [deprecation] AmazonS3EncryptionClient in com.amazonaws.services.s3 has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[24,32] [deprecation] AmazonS3EncryptionClient in com.amazonaws.services.s3 has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[25,38] [deprecation] CryptoConfiguration in com.amazonaws.services.s3.model has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/credentials/provider/factory/strategies/AccessKeyPairCredentialsStrategy.java:[26,29] [deprecation] StaticCredentialsProvider in com.amazonaws.internal has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/wag/InvokeAWSGatewayApi.java:[272,73] [deprecation] getHeaders() in HttpResponse has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/credentials/provider/factory/strategies/AnonymousCredentialsStrategy.java:[44,19] [deprecation] StaticCredentialsProvider in com.amazonaws.internal has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/credentials/provider/factory/strategies/AssumeRoleCredentialsStrategy.java:[163,55] [deprecation] AWSSecurityTokenServiceClient(AWSCredentialsProvider,ClientConfiguration) in AWSSecurityTokenServiceClient has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/credentials/provider/factory/strategies/AssumeRoleCredentialsStrategy.java:[165,32] [deprecation] setEndpoint(String) in AWSSecurityTokenService has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideCEncryptionStrategy.java:[59,8] [deprecation] AmazonS3EncryptionClient in com.amazonaws.services.s3 has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideCEncryptionStrategy.java:[59,46] [deprecation] AmazonS3EncryptionClient in com.amazonaws.services.s3 has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideCEncryptionStrategy.java:[59,42] [deprecation] AmazonS3EncryptionClient(AWSCredentialsProvider,EncryptionMaterialsProvider) in AmazonS3EncryptionClient has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/cloudwatch/PutCloudWatchMetric.java:[251,15] [deprecation] AmazonCloudWatchClient(AWSCredentialsProvider,ClientConfiguration) in AmazonCloudWatchClient has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[51,8] [deprecation] CryptoConfiguration in com.amazonaws.services.s3.model has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[51,47] [deprecation] CryptoConfiguration in com.amazonaws.services.s3.model has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[57,8] [deprecation] AmazonS3EncryptionClient in com.amazonaws.services.s3 has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[57,46] [deprecation] AmazonS3EncryptionClient in com.amazonaws.services.s3 has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/ClientSideKMSEncryptionStrategy.java:[57,42] [deprecation] AmazonS3EncryptionClient(AWSCredentialsProvider,EncryptionMaterialsProvider,CryptoConfiguration) in AmazonS3EncryptionClient has been deprecated
[WARNING] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/credentials/provider/factory/strategies/AccessKeyPairCredentialsStrategy.java:[49,19] [deprecation] StaticCredentialsProvider in com.amazonaws.internal has been deprecated
[INFO] 19 warnings 
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:[486,16] error: onTrigger(ProcessContext,ProcessSessionFactory) in ConsumeKinesisStream cannot override onTrigger(ProcessContext,ProcessSessionFactory) in AbstractProcessor
  overridden method is final
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for nifi-aws-bundle 1.14.0-SNAPSHOT:
[INFO] 
[INFO] nifi-aws-bundle .................................... SUCCESS [  2.902 s]
[INFO] nifi-aws-service-api ............................... SUCCESS [  5.103 s]
[INFO] nifi-aws-abstract-processors ....................... SUCCESS [  8.258 s]
[INFO] nifi-aws-processors ................................ FAILURE [ 30.808 s]
[INFO] nifi-aws-service-api-nar ........................... SUCCESS [ 15.162 s]
[INFO] nifi-aws-nar ....................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  50.014 s (Wall Clock)
[INFO] Finished at: 2021-04-13T06:27:18Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project nifi-aws-processors: Compilation failure
[ERROR] /home/ubuntu/nifi/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:[486,16] error: onTrigger(ProcessContext,ProcessSessionFactory) in ConsumeKinesisStream cannot override onTrigger(ProcessContext,ProcessSessionFactory) in AbstractProcessor
[ERROR]   overridden method is final
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :nifi-aws-processors

```

ChrisSamo632 commented:
@bhaveshpatelh the AbstractProcessor class has been changed as part of this PR (see earlier comments), so you need to build that from this branch in order for the AWS bundle to compile

The easiest (but probably not fastest) thing is probably to just build the whole of nifi

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+            .displayName("Report Metrics to CloudWatch")
+            .name("amazon-kinesis-stream-cloudwatch-flag")
+            .description("Whether to report Kinesis usage metrics to CloudWatch.")
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
```
Should limit allowedValues (true or false)

ChrisSamo632 replied:
Added

pvillard31 commented:
Hey @ChrisSamo632 - It seems there are some build issues with the unit tests.

ChrisSamo632 commented:
> Hey @ChrisSamo632 - It seems there are some build issues with the unit tests.

Yep, my bad - updated the processor and missed the unit test! Should be fixed now @pvillard31 (will keep an eye on the build status)

turcsanyip commented on nifi-api/src/main/java/org/apache/nifi/processor/AbstractProcessor.java:
```diff
@@ -21,7 +21,7 @@
 
     @Override
-    public final void onTrigger(final ProcessContext context, final ProcessSessionFactory sessionFactory) throws ProcessException {
+    public void onTrigger(final ProcessContext context, final ProcessSessionFactory sessionFactory) throws ProcessException {
```
@ChrisSamo632 Thank you for the review changes and apologies for my late reply!

Regarding this particular question of accessing `ProcessSessionFactory` in a processor that comes from `AbstactProcessor` (indirectly):
Basically, we have the following options (as you already mentioned):

1. Remove the processor from the AWS abstract processors hierarchy and extend `AbstractProcessSessionFactory` processor directly (actually, I would not be against it because the code of the AWS abstract processors is quite complicated/deprecated and the new `ConsumeKinesisStream` does not really need the AWS client object provided by the abstract classes)
2. Inherit `AbstractAWSProcessor` from `AbstractSessionFactoryProcessor`, copy this `onTrigger(ProcessContext, ProcessSessionFactory)` implementation (without the `final` modifier) to the AWS base class and override the method where custom handling of `ProcessSessionFactory` is needed (it would be code duplication and I agree with you that it is not really nice)
3. Remove `final` modifier in `AbstractProcessor` (it looks OK to me but I'm a bit uncertain about it because it would affect the base class of almost all processors and I believe it was a design decision to make the `AbstractProcessor`'s API as simple as possible)

@markap14 What is your opinion on option 3? Can the `final` be removed or it should rather be untouched?

ChrisSamo632 replied:
RE option 1 - I considered this but decided against it because the new processor does use *some* of the existing abstract AWS processor logic (e.g. Proxy config & Auth). I didn't much fancy re-implementing that separately and also it could be difficult to keep the refactored processor closely aligned enough to future AWS Processor bundle updates in future (people easily not realising that they needed to make the same changes more than once within the bundle).

I think there should be a separate Jira ticket to refactor the processors (or this could be considered as part of the migration to AWS SDK 2.x, a ticket for which I've previously raised) - the AWS bundle could certainly stand some rework/simplkfication, but this PR is already big enough IMO

I've implemented option 3 currently in this PR, reworking to option 2 shouldn't be a big change if decided that's to preferred way forward.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,680 @@
+
+    @Override
+    protected AmazonKinesisClient createClient(final ProcessContext context, final AWSCredentialsProvider credentialsProvider, final ClientConfiguration config) {
+        awsCredentials = credentialsProvider.getCredentials();
```
The `AWSCredentialsProvider` object is needed later on so it should be stored instead of the `AWSCredentials`.

Furthermore, I would suggest storing as few instance fields as possible in the processor. `AWSCredentialsProvider` can be retrieved directly from `ProcessContext` in `prepareKinesisClientLibConfiguration()`.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,680 @@
+                } catch (Exception e) {
+                    validationResults.add(new ValidationResult.Builder().valid(false)
+                            .subject(STREAM_POSITION_TIMESTAMP.getName())
+                            .explanation(String.format("%s must be parsable by %s", STREAM_POSITION_TIMESTAMP.getDisplayName(),
```
Typo: parseable (?)

ChrisSamo632 replied:
The oh-so-reliable [Wikipedia](https://en.wiktionary.org/wiki/parsable) suggests both versions of the word are correct and "without the e" generally prefered (in formal language at least) - IntelliJ complains at me if I add the `e` so I'll leave as-is unless you really object :slightly_smiling_face:

turcsanyip replied:
Interesting, my IntelliJ is complaining about "parsable" :)
Let's leave it as it is now.

ChrisSamo632 replied:
Good ol' i18n - I'm in UK, guess you're US (or another English speaking locale)?

turcsanyip replied:
Yes, the proofreading language has been set to English (USA) in my IntelliJ.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/record/KinesisRecordProcessorRaw.java:
```diff
@@ -0,0 +1,63 @@
+            log.debug("Sequence No: {}, Partition Key: {}, Data: {}", sequenceNumber, partitionKey, BASE_64_ENCODER.encodeToString(data));
+        }
+
+        session.getProvenanceReporter().receive(flowFile, String.format("http://%s.amazonaws.com/%s/%s#%s", endpointPrefix, kinesisShardId,
```
I'm not familiar with alternative endpoints but my understanding is that a full URL can be configured via the processor's property.
Is it OK to use `.amazonaws.com` suffix in that case?

ChrisSamo632 replied:
I think you're right - the IT tests have an example of using `http://localhost:4566` to override the Kinesis endpoint, so using that as the `transitUri` prefix would make more sense (i.e. where an endpoint override is provided, we should use that, otherwise default to `http://*.amazonaws.com/...`)

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/record/AbstractKinesisRecordProcessor.java:
```diff
@@ -0,0 +1,281 @@
+    String kinesisShardId;
+    long nextCheckpointTimeInMillis;
+
+    boolean processingRecords = false;
```
These fields could be `private` or `protected`.

ChrisSamo632 replied:
Think I'd missed updating some of these from a previous incarnation.

I thought default (package-private) would be better than `protected` because everything that uses these properties are within the same package and I don't (currently) anticipate any classes to extend this abstract class in other packages (without also updating `ConsumeKinesisStream` itself and also then they might as well add `protected` modifier here).

Not `private` because the fields are used by the concrete classes (and unit tests)... but I was being lazy and not generating getter/setter methods, so I've made things `private` and added appropriate getter/setters where needed.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/record/AbstractKinesisRecordProcessor.java:
```diff
@@ -0,0 +1,281 @@
+                processRecordsInput.getRecords().size(), kinesisShardId,
+                processRecordsInput.getCacheEntryTime() != null ? dateTimeFormatter.format(processRecordsInput.getCacheEntryTime().atZone(ZoneId.systemDefault())) : null,
+                processRecordsInput.getCacheExitTime() != null ? dateTimeFormatter.format(processRecordsInput.getCacheExitTime().atZone(ZoneId.systemDefault())) : null,
+                processRecordsInput.getMillisBehindLatest());
```
I don't really like `if (log.isDebugEnabled()) { ... }` but in this case I would add it because `processRecords()` gets called frequently and there are date conversions in the debug call.

ChrisSamo632 replied:
Tried to catch these, but evidently missed one... good spot

turcsanyip reviewed (changes requested):
@ChrisSamo632 Please find my further comments below.

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,680 @@
+            )
+    );
+
+    @SuppressWarnings({"java:S3599", "java:S1171"})
```
They are, I have a `SonarLint` plugin installed in my IDEs and in the habit of fixing/supressing warnings in my normal projects, so just done the same here.

Happy to remove the supressions (or add appropriate comments) if that's preferred - realise adding these will be different to many other parts of the existing code base

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,680 @@
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(
+                    // Kinesis Stream specific properties
+                    KINESIS_STREAM_NAME, APPLICATION_NAME, RECORD_READER, RECORD_WRITER, REGION, ENDPOINT_OVERRIDE,
```
Removed the expression language capability from the property in the abstract class and added a new version to the `PutKinesisStream` processor that adds the expression language capability for just that processor

Could copy & paste the definition into each processor, but sharing the common property definitions seemed sensible still

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+@SystemResourceConsideration(resource = SystemResource.CPU, description = "Kinesis Client Library is used to create a Worker thread for consumption of Kinesis Records. " +
+        "The Worker is initialised and started when this Processor has been triggered. It runs continually, spawning Kinesis Record Processors as required " +
+        "to fetch Kinesis Records. The Worker Thread (and any child Record Processor threads) are not controlled by the normal NiFi scheduler as part of the " +
+        "Concurrent Thread pool are are not released until this processor is stopped.")
```
Typo: "and are not released..." (?)

ChrisSamo632 replied:
D'oh

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+        put("kinesisEndpoint", ENDPOINT_OVERRIDE);
+    }};
+
+    private static final String WORKER_THREAD_NAME_TEMPLATE = ConsumeKinesisStream.class.getName() + "-" + Worker.class.getName() + "-";
```
The thread name with the fully qualified class names is really long. I would consider to use `Class.getSimpleName()` instead.

ChrisSamo632 replied:
Darn, I'd intended to use `getSimpleName()` from the off... good catch!

turcsanyip reviewed (commented):
@ChrisSamo632 Thanks for the review changes!
I will continue with testing the processor with different configurations / advanced cases.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .required(true).build();
+
+    public static final PropertyDescriptor CHECKPOINT_INTERVAL_MILLIS = new PropertyDescriptor.Builder()
```
Similar to "Communications Timeout" property, "Checkpoint Interval" and "Retry Wait" should support other time periods too and it would be enough to convert the period to millisec in the code where it is used.

ChrisSamo632 replied:
Good shout, I only learned about the `asTimePeriod` method after writing this processor and din't think to go back and change things... done

turcsanyip reviewed (commented):
@ChrisSamo632 It seems to me that the checkpoint at shutdown does not work. I have a few uncheckpointed messages that get processed again and again when I restart the processor.

I don't really understand how the graceful shutdown works in KCL, but it seems to me we have no chance to exit cleanly (I am always seeing ZOMBIE state in the record processor's shutdown callback).
Could you please double check it?

```
2021-05-07 08:38:19,416 INFO [NiFi Web Server-476] o.a.n.c.s.StandardProcessScheduler Stopping ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596]
2021-05-07 08:38:19,416 INFO [NiFi Web Server-476] o.a.n.controller.StandardProcessorNode Stopping processor: ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596]
2021-05-07 08:38:19,416 INFO [Timer-Driven Process Thread-4] o.a.n.c.s.TimerDrivenSchedulingAgent Stopped scheduling ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] to run
2021-05-07 08:38:19,416 INFO [Timer-Driven Process Thread-4] o.a.n.p.a.k.stream.ConsumeKinesisStream ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] Requesting Kinesis Worker shutdown
2021-05-07 08:38:19,550 INFO [Flow Service Tasks Thread-2] o.a.nifi.controller.StandardFlowService Saved flow controller org.apache.nifi.controller.FlowController@7d532f2e // Another save pending = false
2021-05-07 08:38:22,801 DEBUG [RecordProcessor-0001] o.a.n.p.a.k.stream.ConsumeKinesisStream ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] Shutting down record processor for shard: shardId-000000000000 with reason: ZOMBIE
2021-05-07 08:38:22,801 DEBUG [RecordProcessor-0000] o.a.n.p.a.k.stream.ConsumeKinesisStream ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] Shutting down record processor for shard: shardId-000000000001 with reason: ZOMBIE
2021-05-07 08:38:22,802 DEBUG [RecordProcessor-0001] o.a.n.p.a.k.stream.ConsumeKinesisStream ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] Checkpointing shard shardId-000000000000
2021-05-07 08:38:22,802 DEBUG [RecordProcessor-0000] o.a.n.p.a.k.stream.ConsumeKinesisStream ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] Checkpointing shard shardId-000000000001
2021-05-07 08:38:22,802 INFO [RecordProcessor-0000] o.a.n.p.a.k.stream.ConsumeKinesisStream ConsumeKinesisStream[id=84a18d0e-9710-3b42-6704-932e41ebe596] Caught shutdown exception, skipping checkpoint.
2021-05-07 08:38:22,802 INFO [RecordProcessor-0000] o.a.n.p.a.k.stream.ConsumeKinesisStream
com.amazonaws.services.kinesis.clientlibrary.exceptions.ShutdownException: Can't update checkpoint - instance doesn't hold the lease for this shard
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.KinesisClientLibLeaseCoordinator.setCheckpoint(KinesisClientLibLeaseCoordinator.java:198)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.RecordProcessorCheckpointer.advancePosition(RecordProcessorCheckpointer.java:307)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.RecordProcessorCheckpointer.checkpoint(RecordProcessorCheckpointer.java:85)
	at org.apache.nifi.processors.aws.kinesis.stream.record.AbstractKinesisRecordProcessor.checkpoint(AbstractKinesisRecordProcessor.java:278)
	at org.apache.nifi.processors.aws.kinesis.stream.record.AbstractKinesisRecordProcessor.checkpointWithRetries(AbstractKinesisRecordProcessor.java:262)
	at org.apache.nifi.processors.aws.kinesis.stream.record.AbstractKinesisRecordProcessor.shutdown(AbstractKinesisRecordProcessor.java:255)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.ShutdownTask.call(ShutdownTask.java:137)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.MetricsCollectingTaskDecorator.call(MetricsCollectingTaskDecorator.java:49)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.MetricsCollectingTaskDecorator.call(MetricsCollectingTaskDecorator.java:24)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Please also find my other comments inline (suggested code refactors / improvements).

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+        final ControllerService recordWriter = validationContext.getProperty(RECORD_WRITER).isSet()
+                ? validationContext.getProperty(RECORD_WRITER).asControllerService()
+                : null;
+        if (recordReader != null && recordWriter == null) {
```
Think I added the booleans after writing this validation logic and didn't think to go back later, good spot

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+    KinesisClientLibConfiguration prepareKinesisClientLibConfiguration(final ProcessContext context, final String appName,
+                                                                       final String streamName, final String workerId,
+                                                                       final String kinesisEndpoint) {
+        final AWSCredentialsProvider credentialsProvider = context.getProperty(AWS_CREDENTIALS_PROVIDER_SERVICE).isSet()
```
The potential problem there would be if someone had a Kinesis stream (AWS or other variant with the same API) that wasn't configured to require security - they'd have no way of using the processor against such a setup

Such a setup is not something that anyone would recommend. Public/anonymous access to a service, particularly one like this, is rarely a good idea... but disallowing it completely in the processor doesn't seem the right thing to do and all other AWS processors allow for anonymous access too (I found the `AnonymousAWSCredentials` use in the `AbstractAWSProcessor` and reproduced it from there as this code now avoids using the Credentials from the inherritted method after a previous suggested refactoring of the code)

turcsanyip replied:
`AWSCredentialsProviderControllerService` has a `Use Anonymous Credentials` option (true/false property). I haven't used it yet but based on its name it should provide the same functionality.

ChrisSamo632 replied:
So it does, you're quite right - and it does, indeed, result in the same thing

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,678 @@
+    protected Collection<ValidationResult> customValidate(final ValidationContext validationContext) {
+        final Set<ValidationResult> validationResults = new HashSet<>(super.customValidate(validationContext));
+
+        initialPositionInStream = InitialPositionInStream.valueOf(StringUtils.trimToEmpty(validationContext.getProperty(INITIAL_STREAM_POSITION).getValue()));
```
Yeah I agree this is better, hadn't checked/realised that Validation & Process Contexts were related with the super class, using the suggested methods is definitely cleaner

ChrisSamo632 commented:
> @ChrisSamo632 It seems to me that the checkpoint at shutdown does not work. I have a few uncheckpointed messages that get processed again and again when I restart the processor.
> 
> I don't really understand how the graceful shutdown works in KCL, but it seems to me we have no chance to exit cleanly (I am always seeing ZOMBIE state in the record processor's shutdown callback).
> Could you please double check it?
> 
> ...

@turcsanyip I'd noticed ZOMBIE logs a few times and thought it seemed a "normal" thing for KCL (from what I could tell, but note I'm hardly a Kinesis/KCL expert). Taking a look around on the internet, there are many threads about people seeing this kind of behaviour in many different versions of the KCL, for example:
* https://github.com/awslabs/amazon-kinesis-client/issues/108
* https://github.com/awslabs/amazon-kinesis-client/issues/677
* https://github.com/amazon-archives/amazon-kinesis-connectors/issues/10

A common thing between these seems to be that the KCL settings likely need tweaking depending upon one's setup. With the ability to configure most KCL settings via Dynamic Properties on the processor, it's largely open for users to figure this out I guess depending upon their setup.

That said, a common theme seems to be the need to increase the `failoverTimeMillis` and/or `gracefulShutdownMillis` KCL settings, I've therefore added those to the `ConsumeKinesisStream` processor's properties to encourage users to investigate/tweak as required.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,708 @@
+            .required(true).build();
+
+    public static final PropertyDescriptor GRACEFUL_SHUTDOWN_PERIOD = new PropertyDescriptor.Builder()
+            .displayName("Failover Time")
```
Copy-paste error: Graceful Shutdown Time
(not sure but "Timeout" may be better, also in case of Failover Timeout)

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,708 @@
+
+    private Date getStartStreamTimestamp(final PropertyContext context, final DateTimeFormatter dateTimeFormatter) {
+        final String streamTimestamp = context.getProperty(STREAM_POSITION_TIMESTAMP).getValue();
+        return new Date(LocalDateTime.parse(streamTimestamp, dateTimeFormatter).toInstant(ZoneOffset.UTC).toEpochMilli());
```
"Stream Position Timestamp" property is expected to be configured in UTC now, which is not really straightforward on the UI because the "aws.kinesis.approximate.arrival.timestamp" FF attribute is written in local TZ (NiFi server's system TZ).
I think local system TZ should be used for this property too (and convert it to UTC for Kinesis in the code) or it should be documented in the property's description that UTC is expected.

turcsanyip replied:
It works properly now with local system time zone. Thanks

A minor note: `withZoneSameInstant(ZoneOffset.UTC)` is not necessarily needed, because in the end we need the Instant and this method does not change the Instant (and for this reason it does not cause any problem either).

ChrisSamo632 replied:
I think I'll leave this as-is because it's a bit more explicit what's going on

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,708 @@
+            }
+        });
+        if (dynamicPropertyFailure.get()) {
+            throw new ProcessException("Failed to set dynamic properties for the Kinesis Client Library (see logs for more details)");
```
Guess I was thinking that this approach would at least log all the issues if there were multiple... but you're right, the `customValidate` should catch any such issues and have invalidated the processor beofre this code could be executed, so no need for this approach and the simpler "throw on first error" can be used

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,708 @@
+        try {
+            if (!PROPERTY_UTILS_BEAN.isWriteable(kclTemp, subject)) {
+                return validationResult
+                        .explanation(String.format("Kinesis Client Library Configuration property with name with%s does not exist or is not writable", StringUtils.capitalize(subject)))
```
It looks good to me now. Thanks!
There is one more "with-" prefix left in `buildDynamicPropertyBeanValidationResult()`.

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/stream/record/TestAbstractKinesisRecordProcessor.java:
```diff
@@ -0,0 +1,272 @@
+
+    private final MockProcessSession session = new MockProcessSession(new SharedSessionState(runner.getProcessor(), new AtomicLong(0)), runner.getProcessor());
+
+    private IRecordProcessor fixture;
```
Probably a hangover from before the abstract class existed and the updates to the tests when I introduced the Record Reader/Writer didn't change the variable type, but it does make sense so easily changed

ChrisSamo632 commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/stream/record/TestAbstractKinesisRecordProcessor.java:
```diff
@@ -0,0 +1,272 @@
+
+        // default test fixture will try operations twice with very little wait in between
+        fixture = new MockKinesisRecordProcessor(processSessionFactory, runner.getLogger(), "kinesis-test",
+                "endpoint-prefix", null, 10_000L, 1L, 2, DATE_TIME_FORMATTER);
```
Think I'd originally expected to use multiple instances of the mock class, but you're right it doesn't make sense any longer so removed

turcsanyip reviewed (commented):
@ChrisSamo632 I found an error case that should be handled in this round in my opinion. If the configured Kinesis Stream does not exist (which is quite a typical user error), then it cannot be seen on the UI. The Kinesis Worker logs error messages but the processor looks fine. After a while, the worker stops but still nothing on the UI.

As far as I see, there is no straightforward way to handle worker errors (eg. pass exception handler in to the worker). `WorkerStateChangeListener` may be used to monitor worker state changes and log error on the processor when the worker stopped for some reason.

turcsanyip commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/ConsumeKinesisStream.java:
```diff
@@ -0,0 +1,726 @@
+            }
+        } else {
+            if (!stopped.get() && WorkerStateChangeListener.WorkerState.SHUT_DOWN == workerState.get()) {
+                throw new ProcessException("Worker has shutdown unexpectedly, possibly due to a configuration issue; check logs for details");
```
`context.yield()` needs to be called before throwing the exception (without yielding `onTrigger()` would be called again immediately and would dump the log with the error message in a loop).

turcsanyip reviewed (approved):
+1 LGTM
Merging to main.

@udaykale Thanks for the initial commit / PR.
@ChrisSamo632 Thanks for the further improvements and getting it across the finish line.

anekar3416 commented:
Just wondering, which version of KCL has this fix or do we still need to use KCL 1.13.3?

ChrisSamo632 commented:
> Just wondering, which version of KCL has this fix or do we still need to use KCL 1.13.3?

@anekar3416 the KCL version was upgraded as part of [NIFI-9520](https://issues.apache.org/jira/browse/NIFI-9520) in PR #5632, released in NiFi 1.16.0. The version of KCL currently used (after the linked ticket & PR) is 1.14.7 - upgrade to KCL 2.x requires larger changes to NiFi first in the future, e.g. NIFI-8287
